cpp_tests/feature/one_hot_encoder_test.cc

```
@@ -122,42 +148,79 @@ TEST_F(OneHotEncoderTest, Mlir) {
+    mlir_inputs.push_back(model_ref);
+    mlir_inputs.push_back(
+        MakeAvailableAsyncValueRef<
+            llvm::ArrayRef<tfrt::RCReference<tfrt::AsyncValue>>>(inputs));
+
+    auto results = model_transform_runner.Run(mlir_inputs);
+    host_context->Await(results);
+    auto outputs =
+        results[0]
+            ->get<llvm::SmallVector<tfrt::RCReference<tfrt::AsyncValue>, 4>>();
+
+    host_context->Await(outputs);
+    SparseVector &actual_vector = outputs[0]->get<SparseVector>();
+    SparseVector expected_vector(2);
+    expected_vector.set(1, 1.0);
+    EXPECT_EQ(actual_vector, expected_vector);
+  }
+
+  {
+    llvm::SmallVector<tfrt::RCReference<tfrt::AsyncValue>, 4> inputs = {
+        MakeAvailableAsyncValueRef<int>(5), MakeAvailableAsyncValueRef<int>(5)};
+
+    llvm::SmallVector<RCReference<AsyncValue>, 4> mlir_inputs;
+    mlir_inputs.push_back(model_ref);
+    mlir_inputs.push_back(
+        MakeAvailableAsyncValueRef<
+            llvm::ArrayRef<tfrt::RCReference<tfrt::AsyncValue>>>(inputs));
 
-  llvm::SmallVector<RCReference<AsyncValue>, 4> invalid_inputs;
-  invalid_inputs.push_back(tfrt::MakeAvailableAsyncValueRef<std::string>(
-      tmp_folder.getAbsolutePath()));
-  invalid_inputs.push_back(tfrt::MakeAvailableAsyncValueRef<int32_t>(5));
-  invalid_inputs.push_back(tfrt::MakeAvailableAsyncValueRef<int32_t>(5));
-
-  auto invalid_results = test::runMlirScript(host_context, mlir_context,
-                                             mlir_script, invalid_inputs);
-  EXPECT_EQ(invalid_results.size(), 1);
-  host_context->Await(invalid_results);
-  EXPECT_TRUE(invalid_results[0]->IsError());
+    auto results = model_transform_runner.Run(mlir_inputs);
+    host_context->Await(results);
+    auto outputs =
+        results[0]
+            ->get<llvm::SmallVector<tfrt::RCReference<tfrt::AsyncValue>, 4>>();
+
+    host_context->Await(outputs);
+    EXPECT_TRUE(outputs[0]->IsError());
```

Could we also check the value of the error message here? This could also helper reader understand why it should return an error.

---

I agree. I'll change it to compare error message strings.

And I'll remove the `EXPECT_TRUE` statement as the comparison above can cover its functionality. If the output happened not to be errors, the generated string would be `"success"`, which cannot equal to any expected error message.

How do you like the idea to create an issue to add a more reasonable log system to Clink's C++ codes? So far we can only get simple error messages like `Column index out of range.`, but I don't know where did this message generate, how to track its stack trace, or how to manage these messages except output to std::out. TFRT has provided tools like `TFRT_RETURN_IF_ERROR` or `DecodedLocation`, but I think there is still space for improvement.

---

Thanks for the explanation. I think it is probably useful to follow the existing C++ logging best practice. If existing C++ logging outputs file name, line number and the error message, we can do the same in Clink.